ClosedXML/Excel/Cells/XLCellsCollection.cs

```diff
@@ -414,5 +424,90 @@ private static void IncrementUsage(Dictionary<int, int> dictionary, Int32 key)
             else
                 dictionary.Add(key, 1);
         }
+
+        private void ChangeColumnNumbers(IDictionary<int, int> columnNumberMappings, Dictionary<int, Dictionary<int, XLCell>> rowsCollection, XLRangeAddress affectedRange)
+        {
+            foreach (var kvp in rowsCollection)
+                ChangeColumnNumbers(columnNumberMappings, kvp.Value, affectedRange);
+        }
+
+        private Dictionary<int, XLCell> ChangeColumnNumbers(IDictionary<int, int> columnNumberMappings, Dictionary<int, XLCell> cellsCollection, XLRangeAddress affectedRange)
+        {
+            var currentColumns = columnNumberMappings
+                .Where(kvp => cellsCollection.TryGetValue(kvp.Key, out var c)
+                              && c.Address.RowNumber >= affectedRange.FirstAddress.RowNumber && c.Address.RowNumber <= affectedRange.LastAddress.RowNumber
+                              && c.Address.ColumnNumber != kvp.Value)
+                .Select(kvp =>
+                {
+                    var c = cellsCollection[kvp.Key];
+                    // Force evaluation of FormulaR1C1 and clear FormulaA1 - this will preserve the relative formula
+                    if (c.HasFormula && !string.IsNullOrWhiteSpace(c.FormulaR1C1)) c.ClearFormulaA1();
+                    c.Address = new XLAddress(c.Worksheet, c.Address.RowNumber, kvp.Value, c.Address.FixedRow, c.Address.FixedColumn);
+
+                    return new
+                    {
+                        OldColumnNumber = kvp.Key,
+                        NewColumnNumber = kvp.Value,
+                        Cell = c
+                    };
+                }).ToArray();
+
+            currentColumns.Select(c => c.OldColumnNumber).ForEach(c => cellsCollection.Remove(c));
+            currentColumns.ForEach(c => cellsCollection.Add(c.NewColumnNumber, c.Cell));
+
+            return cellsCollection;
+        }
+
+        private Dictionary<int, Dictionary<int, XLCell>> ChangeRowNumbers(IDictionary<int, int> rowNumberMappings, Dictionary<int, Dictionary<int, XLCell>> rowsCollection, XLRangeAddress affectedRange)
+        {
+            var currentRows = rowNumberMappings
+                .Select(kvp =>
+                {
+                    var cells = rowsCollection[kvp.Key];
+                    ChangeRowNumbers(rowNumberMappings, cells, affectedRange);
+                    return new
+                    {
+                        OldRowNumber = kvp.Key,
+                        NewRowNumber = kvp.Value,
+                        Cells = cells
+                    };
+                }).ToArray();
```

Comment:
I still have to think about this. I can't imagine that all the tests pass just by accident if there really are side effects.

Reply:
As far as I can tell, the code works correctly.

The only side-effect in the `Select` is modification of cell addresses which doesn't affect rowsCollection directly. 

Maybe the issue was readability? In the code, the `Select`+`ToArray` is used as a `ForEach`. LINQ methods (`Select` among them) should be used as a pure function, i.e. it should produce data based on inputs and not mutate anything (IEnumerable chain can be evaluated multiple times). That is why `ForEach` in most implementations around is a terminal node of IEnumerable chain (returns void instead of IEnumerable) - to make sure evaluation of sideeffects happens only once.

As noted, the loop wouldn't work, even simple switch of two rows 1->2, 2->1 would end up with a lost row).